retrybp/retry.go

```diff
@@ -0,0 +1,60 @@
+package retrybp
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	retry "github.com/avast/retry-go"
+
+	"github.com/reddit/baseplate.go/batcherror"
+)
+
+func init() {
+	retry.DefaultAttempts = 1
+	retry.DefaultDelay = 1 * time.Millisecond
+	retry.DefaultMaxJitter = 5 * time.Millisecond
+	retry.DefaultDelayType = retry.CombineDelay(retry.BackOffDelay, retry.RandomDelay)
+	retry.DefaultLastErrorOnly = false
+}
+
+type contextKeyType struct{}
+
+var contextKey contextKeyType
+
+// WithOptions sets the given retry.Options on the given context.
+func WithOptions(ctx context.Context, options ...retry.Option) context.Context {
+	return context.WithValue(ctx, contextKey, options)
+}
+
+// GetOptions returns the list of retry.Options set on the context.
+func GetOptions(ctx context.Context) (options []retry.Option, ok bool) {
+	options, ok = ctx.Value(contextKey).([]retry.Option)
+	return
+}
+
+// Do retries the given function using retry.Do with the default retry.Options
+// provided and overriding them with any options set on the context via
+// WithOptions.
+//
+// The changes this has compared to retry.Do are:
+//
+// 1. Pulling options from the context.  This allows it to be used in middleware
```

Comment:
"Use context Values only for request-scoped data that transits processes and APIs, not for passing optional parameters to functions."

🔕 We might be stretching this limit.

Reply:
> We might be stretching this limit.

Fair, although this is a fairly common way to stretch the limit, this is _really_ only for when you are using retry middleware though where you don't really have another option.

Reply:
I am not sure if this is actually a common way, other libraries might misbehave in that case too. Usually this calls for a separate type.

Reply:
The main one I'm thinking of is Thrift where you have to use the context object to pass through headers.

I'm not sure there's a good way around using it this way if we want to have retry middleware (so you can have default retries plus customization).  One option would be to just _remove_ the retry middleware and explicitly use `retrybp.Do` each time?